renderer/src/shared/Hint/positioning.ts

```diff
@@ -0,0 +1,278 @@
+import { computePosition } from "@floating-ui/dom";
+import { Dispatch, MutableRefObject, SetStateAction } from "react";
+import { Position } from "./types";
+
+export const positioning = (
+  learnHint: () => void,
+  referenceRef: MutableRefObject<any>,
+  floatingRef: MutableRefObject<any>,
+  SVGRef: MutableRefObject<any>,
+  learningMethod: "buttonClick" | "nextClick" | "ctrlArrowNumberPress",
+  position: Position,
+  showHint: boolean,
+  shiftY: number,
+  shiftX: number,
+  HorizontalLineRef: MutableRefObject<any>,
+  VerticalLineRef: MutableRefObject<any>,
+  TriangleRef: MutableRefObject<any>,
+  setShowHint: Dispatch<SetStateAction<boolean>>
+): void => {
+  const handleKeyDown = (e) => {
+    if (
+      (e.ctrlKey && e.key === "ArrowUp") ||
+      (e.key === "Meta" && e.key === "ArrowUp") ||
+      ((e.ctrlKey || e.key === "Control" || e.key === "Meta") &&
+        /^[0-9]$/.test(e.key))
+    ) {
+      learnHint();
+    }
+  };
+  if (referenceRef && floatingRef.current && SVGRef.current) {
+    const hintHeight = floatingRef.current.offsetHeight;
+    const hintWidth = floatingRef.current.offsetWidth;
+    const HorizontalLine = HorizontalLineRef.current;
+    const VerticalLine = VerticalLineRef.current;
+
+    switch (learningMethod) {
+      case "buttonClick":
+        referenceRef.current.addEventListener("click", () => {
+          learnHint();
+        });
+        break;
+
+      case "ctrlArrowNumberPress":
+        window.addEventListener("keydown", handleKeyDown);
+        break;
+    }
+
+    computePosition(referenceRef.current, floatingRef.current, {
+      placement: position.basePosition,
+    })
+      .then(({ x, y }) => {
+        if (showHint) {
+          Object.assign(referenceRef.current.style, {
+            "z-index": "50",
+            position: "relative",
+          });
+        } else {
+          Object.assign(referenceRef.current.style, {
+            "z-index": null,
+          });
+        }
+
+        switch (position.basePosition) {
+          case "top":
+            Object.assign(floatingRef.current.style, {
```

Comment:
Could you please move this logic to the separate utils function?
From 65 line to 120.